### ISSUE TITLE
Different accuracy with identical models [ONNX, Accuracy]

### ISSUE THREAD

**Issue by ivkalgin** [NVIDIA/TensorRT#2182]
## Description

We got different accuracy for identical onnx models. Models have identical graphs and parameters. Nodes of both models are topologically sorted but order of nodes in files is different.

Result with TRT 8.4.1.5 + RTX3080ti:

| build   | accuracy |
|---------|----------|
| model_1 | 57.04%   |
| model_2 | 61.74%   |

Expected result:

Accuracy of model_1 should be equal accuracy of model_2.

## Environment

**TensorRT Version**: 8.4.1.5
**NVIDIA GPU**: tested on RTX 3080ti and a6000 
**NVIDIA Driver Version**: 510.73.08
**CUDA Version**: 11.7.60
**CUDNN Version**: 8.4.0.27
**Operating System**: Ubuntu 20.04.3 LTS
**Python Version (if applicable)**: 3.8
**PyTorch Version (if applicable)**: 1.11

## Steps To Reproduce

Full instruction can be found [here](https://github.com/ivkalgin/TRT_84_accuracy_issue).

**Comment by zerollzeng** [comment]
May I ask what's the purpose of the two models with the same nodes but in different orders? I just think the order of computing surely will affect the final result. e.g. (a+b)*c != a*c+b

BTW The 23GB model size is terrifying.

**Comment by ivkalgin** [comment]
I try to quantize GPTJ for question answering task. After onnx optimization I always got the same optimized model, but order of the nodes is different from run to run. My experiments with a6000+trt 8.4 showed that final accuracy varies randomly from 60% to 90%, but experiments with **a6000+trt 8.2 showed stable result** 92%. 92% agrees perfectly with the quantized torch model.
PS
It is GPTJ so 23GB it is ok ).

**Comment by ivkalgin** [comment]
> May I ask what's the purpose of the two models with the same nodes but in different orders? I just think the order of computing surely will affect the final result. e.g. (a+b)_c != a_c+b

I try to quantize GPTJ for question answering task. After onnx optimization I always got the same optimized model, but order of the nodes is different from run to run. My experiments with a6000+trt 8.4 showed that final accuracy varies randomly from 60% to 90%, but experiments with a6000+trt 8.2 showed stable result 92%. 92% agrees perfectly with the quantized torch model.

> BTW The 23GB model size is terrifying.

It is GPTJ so 23GB it is ok ).

**Comment by ivkalgin** [comment]
Hello @zerollzeng 

Did you reproduce the bug? If you have any problems with downloading the model, I can upload it to any resource that you specify.

PS
I made a mistake in a previous post.
> I try to quantize GPTJ for question answering task.
"question answering" -> "sentence classification"

**Comment by zerollzeng** [comment]
> After onnx optimization I always got the same optimized model, but order of the nodes is different from run to run. 

I can see there are 2 onnx in your reproduce repo. so they are not the same right? 

> My experiments with a6000+trt 8.4 showed that final accuracy varies randomly from 60% to 90%, but experiments with a6000+trt 8.2 showed stable result 92%. 92% agrees perfectly with the quantized torch model.

Do you mean you got different results on the same model? e.g. on the first test you got 50%, but in the second test you got 60%?

**Comment by ivkalgin** [comment]
> > After onnx optimization I always got the same optimized model, but order of the nodes is different from run to run.
> 
> I can see there are 2 onnx in your reproduce repo. so they are not the same right?
> 
 Yes, it is different files, but models are the same in terms of onnx specification (have the same graphs). They have only one difference: the order of onnx nodes in file.

**Comment by ivkalgin** [comment]
> Do you mean you got different results on the same model? e.g. on the first test you got 50%, but in the second test you got 60%?

No, I got different results with different files (model_1,onnx, model_2.onnx), but in terms of onnx specification it is identical models.

**Comment by ivkalgin** [comment]
I try to clarify difference between model_1.onnx and model_2.onnx

Imagen simple graph that contains only 3 nodes: 
A -- C--> 
B --/
In terms of onnx specification you can write list of nodes in 2 ways:
1. [ABC]
2. [BAC]
Both variants are valid, because both topologically sorted. Inference of both variants must return the same results.

**Comment by zerollzeng** [comment]
@kevinch-nv @ttyio @pranavm-nvidia Can you help on this issue?

**Comment by ivkalgin** [comment]
Hello @zerollzeng,
Do you have new information about this bug?

**Comment by zerollzeng** [comment]
I've filed internal bug 3754510 to track this.

**Comment by zerollzeng** [comment]
Hi @ivkalgin can you upload the weight file and onnx model to Google Drive?

**Comment by ivkalgin** [comment]
Hi @zerollzeng .
Please let me know when your are downloaded all files. I should release the disk by August 26th.
[model_1.onnx](https://drive.google.com/file/d/13lbl2H_rVobczB9bvOuG74nNvkNJsyTo/view?usp=sharing)
[model_2.onnx](https://drive.google.com/file/d/1cbC_UNp5c2m_D6fdRhDxdRoEKtVFvvfV/view?usp=sharing)
[model.onnx.data](https://drive.google.com/file/d/1CnOVssh8pw5mX6rUoCe9vd7zbKTbwE6Q/view?usp=sharing)

**Comment by ivkalgin** [comment]
Hello @zerollzeng,
Did you download models? Can I remove data from google drive?

**Comment by zerollzeng** [comment]
Yes, we've downloaded it.

**Comment by zerollzeng** [comment]
> I try to quantize GPTJ for question answering task. After onnx optimization I always got the same optimized model, but order of the nodes is different from run to run. My experiments with a6000+trt 8.4 showed that final accuracy varies randomly from 60% to 90%, but experiments with a6000+trt 8.2 showed stable result 92%. 92% agrees perfectly with the quantized torch model.

Hi @ivkalgin,
using 8.2.5.2, we see that the two produced engines have different accuracy, using your validation script:
model_1: ACCURACY: {'accuracy': 0.7332667332667333}
model_2: ACCURACY: {'accuracy': 0.7302697302697303}

Do you have any idea why we can't reproduce the 92% accuracy?

**Comment by ivkalgin** [comment]
> > I try to quantize GPTJ for question answering task. After onnx optimization I always got the same optimized model, but order of the nodes is different from run to run. My experiments with a6000+trt 8.4 showed that final accuracy varies randomly from 60% to 90%, but experiments with a6000+trt 8.2 showed stable result 92%. 92% agrees perfectly with the quantized torch model.
> 
> Hi @ivkalgin, using 8.2.5.2, we see that the two produced engines have different accuracy, using your validation script: model_1: ACCURACY: {'accuracy': 0.7332667332667333} model_2: ACCURACY: {'accuracy': 0.7302697302697303}
> 
> Do you have any idea why we can't reproduce the 92% accuracy? what's the specific version of TRT 8.2 you are using?

Hello @zerollzeng,
What GPU do you use? 92% it is my result on a6000. I think result depends on GPU model and maybe depends on CPU.

**Comment by netaz** [comment]
@ivkalgin you should not be seeing 92% accuracy on one GPU and 73% on a different machine unless something it broken. Obviously, it could be a problem with TRT but if you are saying that the accuracy is unstable when you (1) change GPU; or (2) make small changes to the order of ONNX nodes (keeping same processing semantics); or (3) when we change TRT version - then I think something could be wrong with the model.

Can you explain how you trained this model?
* Did you start with a pretrained model w/o QAT? If yes, does the FP32 model (unquantized) also shows instability?
* How did you add the QDQ nodes and how did you determine the scales (what SW did you use? did you perform calibration? Was the calibration DS large enough?)?
* Did you perform fine-tuning after adding fake quantization? Did you observe the loss vs accuracy curve? Did you check that you did not overfit?

Intuitively I think you should verify that your model is not overfitting because an overfitted model will be unstable when we introduce noise from quantization and limited-precision arithmetic (in float arithmetic different operations ordering can produce small differences in output).

**Comment by netaz** [comment]
@ivkalgin you should not be seeing 92% accuracy on one GPU and 73% on a different machine unless something it broken. Obviously, it could be a problem with TRT but if you are saying that the accuracy is unstable when you (1) change GPU; or (2) make small changes to the order of ONNX nodes (keeping same processing semantics); or (3) when we change TRT version - then I think something could be wrong with the model.

Can you explain how you trained this model?

Did you start with a pretrained model w/o QAT? If yes, does the FP32 model (unquantized) also shows instability?
How did you add the QDQ nodes and how did you determine the scales (what SW did you use? did you perform calibration? Was the calibration DS large enough?)?
Did you perform fine-tuning after adding fake quantization? Did you observe the loss vs accuracy curve? Did you check that you did not overfit?
Intuitively I think you should verify that your model is not overfitting because an overfitted model will be unstable when we introduce noise from quantization and limited-precision arithmetic (in float arithmetic different operations ordering can produce small differences in output).

**Comment by nzmora-nvidia** [comment]
@ivkalgin you should not be seeing 92% accuracy on one GPU and 73% on a different machine unless something it broken. Obviously, it could be a problem with TRT but if you are saying that the accuracy is unstable when you (1) change GPU; or (2) make small changes to the order of ONNX nodes (keeping same processing semantics); or (3) when we change TRT version - then I think something could be wrong with the model.

Can you explain how you trained this model?

Did you start with a pretrained model w/o QAT? If yes, does the FP32 model (unquantized) also shows instability?
How did you add the QDQ nodes and how did you determine the scales (what SW did you use? did you perform calibration? Was the calibration DS large enough?)?
Did you perform fine-tuning after adding fake quantization? Did you observe the loss vs accuracy curve? Did you check that you did not overfit?
Intuitively I think you should verify that your model is not overfitting because an overfitted model will be unstable when we introduce noise from quantization and limited-precision arithmetic (in float arithmetic different operations ordering can produce small differences in output).

**Comment by zerollzeng** [comment]
This issue is fixed in 8.5.1. please wait for it to go public. closed.
